src/sentry/utils/snuba.py

```diff
@@ -0,0 +1,162 @@
+    issues = get_project_issues(project_ids) if references_issues else None
+
+    url = '{0}/query'.format(SNUBA)
+    request = {k: v for k, v in six.iteritems({
+        'from_date': start.isoformat(),
+        'to_date': end.isoformat(),
+        'conditions': conditions,
+        'groupby': groupby,
+        'project': project_ids,
+        'aggregation': aggregation,
+        'aggregateby': aggregateby,
+        'granularity': rollup,
+        'issues': issues,
+        'arrayjoin': arrayjoin,
+    }) if v is not None}
+
+    response = requests.post(url, data=json.dumps(request))
+    # TODO handle error responses
+    response = json.loads(response.text)
+
+    # Validate and scrub response, and translate snuba keys back to IDs
+    expected_cols = groupby + ['aggregate']
+    assert all(c['name'] in expected_cols for c in response['meta'])
+    for d in response['data']:
+        if 'time' in d:
+            d['time'] = int(to_timestamp(parse_datetime(d['time'])))
+        if d['aggregate'] is None:
+            d['aggregate'] = 0
+        for col in rev_snuba_map:
+            if col in d:
+                d[col] = rev_snuba_map[col][d[col]]
+
+    return nest_groups(response['data'], groupby)
+
+
+def nest_groups(data, groups):
+    """
+    Build a nested mapping from query response rows. Each group column
+    gives a new level of nesting and the leaf result is the aggregate
+    """
+    if not groups:
+        # If no groups, just return the aggregate value from the first row
+        return data[0]['aggregate'] if data else None
+    else:
+        g, rest = groups[0], groups[1:]
+        inter = {}
+        for d in data:
+            inter.setdefault(d[g], []).append(d)
+        return {k: nest_groups(v, rest) for k, v in six.iteritems(inter)}
+
+# The following are functions for resolving information from sentry models
+# about projects, environments, and issues (groups). Having this snuba
+# implementation have to know about these relationships is not ideal, and
+# many of these relationships (eg environment id->name) will have already
+# been queried and exist somewhere in the call stack, but for now, lookup
+# is implemented here for simplicity.
+
+
+def get_snuba_map(column, ids):
+    """
+    Some models are stored differently in snuba, eg. as the environment
+    name instead of the the environment ID. Here we look up a set of keys
+    for a given model and return a lookup dictionary from those keys to the
+    equivalent ones in snuba.
+    """
+    mappings = {
+        'environment': (Environment, 'name'),
+        'release': (Release, 'version'),
+    }
+    if column in mappings and ids:
+        model, field = mappings[column]
+        return dict(model.objects.filter(id__in=ids).values_list('id', field))
+    return None
+
+
+def get_project_issues(project_ids):
+    """
+    Get a list of issues and associated fingerprint hashes for a project.
+    """
```

Comment:
This is returning every hash a project has? That definitely won't fit in RAM or in a query, and what's the difference between doing `primary_hash IN (every_hash_in_the_project)` and `project_id = FOO`?

Doesn't this method need to take a `group_id` argument and only return those hashes?

Reply:
Hmm, this could be a problem. In the case where you are only looking at a single group, then we should reduce the list here for sure. The [snuba](https://github.com/getsentry/snuba/blob/master/snuba/util.py#L43) code already does reduce the issue expression to only the ones that are needed to filter/group by, but I guess its better to filter here so we don;t have to send as many hashes to snuba.

The thing is though, with things like `TSDBModel.frequent_issues_by_project` which would require sending all the issue->hash mappings, so that we could compute the most frequently seen issues. If that list is not going to fit in memory I guess we have a problem

Reply:
Just to repeat what I said in #snuba, my only idea is to `select primary_hash, count() as c from sentry_dist where project_id = X group by primary_hash order by c desc limit 100` and then on Sentry side we'd have to reconcile the `(hash, count)` results into the biggest groups.

I definitely think "send all project hashes in the query" won't work, though. I could be wrong!